grudge/discretization.py

```diff
@@ -636,17 +563,166 @@ def normal(self, dd):
+
+
+# {{{ Discretization construction routines
+
+def make_discretization_collection(
+        array_context: ArrayContext,
+        discr_context,
+        order=None,
+        discr_tag_to_group_factory=None,
+        mpi_communicator=None) -> DiscretizationCollection:
+    """Construct a discretization collection given an array context *array_context*
+    and user-provided discretization parameters.
+
+    :arg discr_context: A :class:`~meshmode.mesh.Mesh` object to define
+        the discretization collection over.
+    :arg discr_tag_to_group_factory: A mapping from discretization tags
+        (typically one of: :class:`grudge.dof_desc.DISCR_TAG_BASE`,
+        :class:`grudge.dof_desc.DISCR_TAG_MODAL`, or
+        :class:`grudge.dof_desc.DISCR_TAG_QUAD`) to a
+        :class:`~meshmode.discretization.poly_element.ElementGroupFactory`
+        indicating with which type of discretization the operations are
+        to be carried out, or *None* to indicate that operations with this
+        discretization tag should be carried out with the standard volume
+        discretization.
+    :arg mpi_communicator: An (optional) MPI communicator.
+    :returns: A :class:`DiscretizationCollection`.
+    """
+    if not isinstance(discr_context, Mesh):
+        raise NotImplementedError(
+            "Currently `discr_context` must be a meshmode Mesh object"
+        )
+    mesh = discr_context
 
-class DGDiscretizationWithBoundaries(DiscretizationCollection):
-    def __init__(self, *args, **kwargs):
-        warn("DGDiscretizationWithBoundaries is deprecated and will go away "
-                "in 2022. Use DiscretizationCollection instead.",
-                DeprecationWarning, stacklevel=2)
+    from meshmode.discretization.poly_element import \
+        default_simplex_group_factory
 
-        super().__init__(*args, **kwargs)
+    if discr_tag_to_group_factory is None:
+        if order is None:
+            raise TypeError(
+                "one of 'order' and 'discr_tag_to_group_factory' must be given"
+            )
+
+        discr_tag_to_group_factory = {
+            DISCR_TAG_BASE: default_simplex_group_factory(base_dim=mesh.dim,
+                                                          order=order)
+        }
+    else:
+        if order is not None:
+            discr_tag_to_group_factory = discr_tag_to_group_factory.copy()
+            if DISCR_TAG_BASE in discr_tag_to_group_factory:
+                raise ValueError(
+                    "if 'order' is given, 'discr_tag_to_group_factory' must "
+                    "not have a key of DISCR_TAG_BASE"
+                )
+
+            discr_tag_to_group_factory[DISCR_TAG_BASE] = \
+                default_simplex_group_factory(base_dim=mesh.dim, order=order)
+
+    # Modal discr should always comes from the base discretization
+    discr_tag_to_group_factory[DISCR_TAG_MODAL] = \
+        _generate_modal_group_factory(
+            discr_tag_to_group_factory[DISCR_TAG_BASE]
+        )
```

Comment:
This should error if `DISCR_TAG_MODAL` was supplied, instead of silently overwriting. The docs specifically allow supplying modal.

Reply:
I've updated this in https://github.com/inducer/grudge/pull/122/commits/fa9a8b904134bd46c7ed6ceef4b656d98c76d467; I've opted to only provide the group factor when it's not provided by the user